README.md

```diff
@@ -561,52 +619,85 @@ class MyBaseClass extends WidgetBase<WidgetProperties> {
+
+Create an `Injector` instance and pass the `payload` that needs to be injected to the constructor:
 
 ```ts
-class MyBaseClass extends WidgetBase<WidgetProperties> {
-    @afterRender()
-    myAfterRender(result: DNode): DNode {
-        // do something with the result
-        return result;
-    }
+const injector = new Injector({ foo: 'baz' });
+registry.defineInjector('my-injector', injector);
+```
+
+To connect the registered `injector` to a widget we can use the `Container` HOC (higher order component) provided by `widget-core`. The `Container` accepts a widget `constructor`, `injector` label and `getProperties` mapping function as arguments and returns a new class that returns the passed widget from its `render` function.
+
+`getProperties` receives the `payload` from an `injector` and `properties` from the container HOC component, that are used to map into the wrapped widgets properties.
+
+```ts
+import { Container } from '@dojo/widget-core/Container';
+import { MyWidget } from './MyWidget';
+
+function getProperties(payload: any, properties: any) {
+    return {
+        foo: payload.foo
+    };
 }
+
+export const MyWidgetContainer = Container(MyWidget, 'my-injector', getProperties);
 ```
 
-And using the `afterRender` function for non decorator environments:
+The returned class from `Container` HOC is then be used in place widget it wraps, not although the container assumes the properties type of the wrapped widget, they all become optional.
 
 ```ts
-class MyBaseClass extends WidgetBase<WidgetProperties> {
-    constructor() {
-        super();
-        afterRender(this.myOtherAfterRender)(this);
-    }
+// import { MyWidget } from './MyWidget';
+import { MyWidgetContainer } from './MyWidgetContainer';
 
-    myOtherAfterRender(result: DNode): DNode {
-        // do something with the result
-        return result;
+interface AppProperties {
+    foo: string;
+}
+
+class App extends WidgetBase<AppProperties> {
+    render() {
+        return v('div', {}, [
+            // w(MyWidget, { foo: 'bar' })
+            w(MyWidgetContainer, {})
+        ]);
     }
 }
 ```
 
+**Please Note:** Don't use container as the top level application widget, containers always `render` which means that you'll find your application is calling `render` every request animation frame even if there is no need for the `render`.
+
+### Decorators
+
+All built-in decorators can be used in non-decorator environments (Either JavaScript/ES6 or a TypeScript project that does not have the experimental decorators configuration set to true in the `tsconfig`) programmatically calling them directly, usually within a Widget classes `constructor`.
```

Comment:
we use builtin, built in, and built-in, we should decide on one and be consistent